### PR TITLE
Added option to select the directory for saving the png file.

### LINKED ISSUE #130: Being able to generate burndown chart image
Currently we have an option to select the working directory, where the data file is. At the moment the `png` with the burndown chart is in the same directory. We should implement a new option in `burndown` command, so that this directory can be different.

### PR DESCRIPTION
**Pull request by rishabhptr**
Fixes #130

### CODE REVIEW

**Comment by rishabhptr** [comment]
@Ana06 Yes you are right :smile: . Actually it is because when we execute the python scripts we change directory to the output directory so I think if we store the current directory  user would be able to choose path relative to his current path. I have made the change.

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,8 @@
 
 ## Version 0.2.0
 
+* Allow to select location to save burndown chart file through `--location`
```
Comment (Ana06) [inline CHANGELOG.md]: I would say **image** instead of file, to avoid confusion with the `.yml` file.

File: scripts/create_burndown.py [hunk 2]
```diff
@@ -26,12 +26,16 @@ def parseCommandLine():
...
   parser.add_argument('--output', help='Location of data to process')
+  parser.add_argument('--location', help='Image file location')
   parser.add_argument('--no-tasks', action='store_true', help='Disable Tasks line in the chart', default=False)
   parser.add_argument('--with-fast-lane', action='store_true', help='Draw line for Fast Lane cards', default=False)
   parser.add_argument('--verbose', action='store_true', help='Verbose Output', default=False)
   parser.add_argument('--no-head', action='store_true', help='Run in headless mode', default=False)
   args = parser.parse_args()
 
+  if args.location:
+    args.location = os.path.join(os.getcwd(),args.location)
```
Comment (Ana06) [inline scripts/create_burndown.py]: what about doing this inside `saveImage`? It is actually not needed at this point. 😉

File: scripts/plot.py [hunk 3]
```diff
@@ -68,7 +69,15 @@ def drawLegend(self, legend_list):
     return
 
   def saveImage(self, args):
-    plt.savefig('burndown-' + args.sprint + '.png',bbox_inches='tight')
+    path = args.location
+    filename = 'burndown' + args.sprint + '.png'
+    if path:
```
Comment (Ana06) [inline scripts/plot.py]: you could write

```python
if args.location:
  path = os.path.join(os.getcwd(),args.location)
```

File: scripts/create_burndown.py (hunk 2 above)
Comment (rishabhptr) [reply scripts/create_burndown.py]: Actually, when we pass --output=' ' for indicating the data file we switch to that directory at the very next line of code and saveImage is called in the last so when saveImage is called the current directory would be 'output' directory and not the directory the user is in now so I updated this here. :smile:

**Comment by rishabhptr** [comment]
@Ana06 I have added the test but I encountered a issue because as with current functionality we open the image file automatically upon creation which would occur in running test too which isn't desired.

I noticed there is a option '--no-head' for this specific issue but it wasn't working as it was not present in cli.rb so I added it there and using it in my test.Please suggest if I should make some change in this.

File: lib/burndown_chart.rb [hunk 4]
```diff
@@ -160,6 +160,7 @@ def process_options(hash)
         cli_switches << "--output #{hash['output']}" if hash['output']
         cli_switches << '--verbose'                  if hash['verbose']
         cli_switches << "--location #{hash['location']}" if hash['location']
+        cli_switches << '--no-head' if hash['no-head']
```
Comment (Ana06) [inline lib/burndown_chart.rb]: you shouldn't modify the code to be able to write a test 😱

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,8 @@
 
 ## Version 0.2.0
 
+* Allow to select location to save burndown chart image through `--location`
```
Comment (Ana06) [inline CHANGELOG.md]: this should have been in the _Master (unreleased)_ section, but we have decided in https://github.com/openSUSE/trollolo/pull/163 to stop using this. So please remove this, it is not needed any more. And sorry for the mess. 😄

File: lib/burndown_chart.rb (hunk 4 above)
Comment (rishabhptr) [reply lib/burndown_chart.rb]: Sorry, actually I thought this could be implemented as a option and didnt realize that it had to be passed directly to python scripts for automation.

**Comment by rishabhptr** [comment]
@Ana06 Thanks for the review, I have made some changes. I have re-written test to directly use no-head by passing it while executing python script.

**Comment by Ana06** [comment]
@rishabhptr the number of contributors in Trollolo is increasing what is decreasing the velocity to review PR 😞  I'll take a look at soon as possible. 😉

**Comment by rishabhptr** [comment]
@Ana06 No problem 😄 , and actually there is more work to it as earlier travis was showing error for not having python-matplotlib installed so I think we might need to add it in travis and also set up some display parameters to successfully execute the python script as directory and file will be created by them.

**Comment by rishabhptr** [comment]
@Ana06 The tests are running fine locally, but on travis I think they need matplot installed to run the script so I think a change in travis configuration might be needed to test it properly.

File: .travis.yml [hunk 6]
```diff
@@ -3,6 +3,9 @@ cache: bundler
 before_install:
   - gem --version
   - gem list bundler
+  - sudo apt-get install python-matplotlib
+  - "export DISPLAY=:99.0"
```
Comment (Ana06) [inline .travis.yml]: why is this needed? :confused:

File: .travis.yml [hunk 7]
```diff
@@ -3,6 +3,9 @@ cache: bundler
...
+  - "sh -e /etc/init.d/xvfb start"
```
Comment (Ana06) [inline .travis.yml]: why is this needed? :confused:

File: spec/unit/burndown_chart_spec.rb [hunk 8]
```diff
@@ -492,12 +492,21 @@
...
+
     it 'sends joined parsed options to python script' do
       allow(described_class).to receive(:process_options).and_return(%w{ --test 1 --no-blah })
       allow(described_class).to receive(:plot_helper).and_return('mescript')
       expect(described_class).to receive(:system).with('python mescript 42 --test 1 --no-blah')
       described_class.plot(42, foo: 1, bar: 2)
     end
+
+    it 'saves image file in given location' do
+      dir = given_directory
```
Comment (Ana06) [inline spec/unit/burndown_chart_spec.rb]: why do you need this variable? :confused:

File: spec/unit/burndown_chart_spec.rb [hunk 9]
```diff
@@ -492,12 +492,21 @@
...
+      allow(described_class).to receive(:process_options).and_return(%W{--output spec/data/burndown_dir --location #{dir} --no-head})
+      described_class.plot(2, foo: 1)
+      expect(File.exist?(File.join(dir, 'burndown02.png'))).to be true
```
Comment (Ana06) [inline spec/unit/burndown_chart_spec.rb]: does this mean that test is creating the Image? Should we maybe delete it? I am not sure if that happens with other tests. :thinking:

File: spec/unit/burndown_chart_spec.rb (hunk 8 above)
Comment (rishabhptr) [reply spec/unit/burndown_chart_spec.rb]: It is actually the temporary directory created for writing the test data ( .png file ) ,as  we are using GivenFileSystemSpecHelpers in tests. 
http://www.rubydoc.info/gems/given_filesystem/0.1.2/file/README.md#Thanks

File: spec/unit/burndown_chart_spec.rb (hunk 9 above)
Comment (rishabhptr) [reply spec/unit/burndown_chart_spec.rb]: Yes, we are creating the image from our test data that already exists because then only we can test that image would be saved in the right place.
As ,I am mocking the parameters to point '--location' to the temp directory I dont think we need to do anything else as file wont exist after test run.

Also, I did this because this is how we are testing the functionality of '--output' option to  check that data file (.YAML) are saved in correct directory.

https://github.com/openSUSE/trollolo/blob/4f2df0497e6fa6e168eed7cc17d849335312ccda/spec/unit/cli_spec.rb#L14-L20

File: .travis.yml (hunk 7 above)
Comment (rishabhptr) [reply .travis.yml]: I had run travis on my fork and had faced issue of not having any display parameter for running python script so I used this here for headless GUI for testing.
https://docs.travis-ci.com/user/gui-and-headless-browsers/

Comment (cornelius) [reply .travis.yml]: Does that mean that you are running matplotlib in the normal unit tests? I would avoid that because the setup can be tricky and strongly depends on the environment. This doesn't give results well defined enough to be reliably checked in tests.

We have the integration tests which run matplotlib in a container: `integration/create_burndown_spec.rb` This works very reliably.

File: spec/unit/burndown_chart_spec.rb (hunk 8 above)
Comment (Ana06) [reply spec/unit/burndown_chart_spec.rb]: but why do you need to assign it to a variable? :confused:

File: .travis.yml (hunk 7 above)
Comment (Ana06) [reply .travis.yml]: I agree with @cornelius :+1:

Comment (rishabhptr) [reply .travis.yml]: Yes if we have a docker container for matplotlib then it should be used,I will make the changes and also sorry for late response as I was busy with exams past week. :smile:

**Comment by Ana06** [comment]
@rishabhptr what is the state of this? do you need help? :smile:

**Comment by rishabhptr** [comment]
@Ana06  As I mentioned earlier in comments I am sorry for late response due to exams. Also if we have docker container for matplotlib it was not required in travis :smile:  so I have made the changes.

**Comment by rishabhptr** [comment]
@Ana06 Thanks for the reply :smile: . Sorry for the late response as I was away, I have rebased my branch but I'm encountering a syntax error in one of the python files called by --plot command. I am not really sure if it's same for others.

**Comment by rishabhptr** [comment]
@Ana06 Yes the error has now resolved after rebase :smiley: . I have resolved the conflicts, please have a look.

File: scripts/plot.py [hunk 10]
```diff
@@ -68,9 +69,17 @@ def drawLegend(self, legend_list):
     return
 
   def saveImage(self, args):
-    plt.savefig('burndown-' + args.sprint + '.png',bbox_inches='tight')
+    path = args.location
+    filename = 'burndown-' + args.sprint + '.png'
```
Comment (Ana06) [inline scripts/plot.py]: I would name this variable path_image and add the path to it, so that you don't need the else, so:
```ruby
path = args.location
image_path = 'burndown-' + args.sprint + '.png'

if path
  if not os.path.exists(path):
    os.makedirs(path)
  image_path = os.path.join(path, image_path)

plt.savefig(image_path, bbox_inches='tight')

Comment (rishabhptr) [reply scripts/plot.py]: Yeah this cleans up some code, thanks for the suggestion :smile:

**Comment by rishabhptr** [comment]
@Ana06 Thanks for the review, I have made the changes :smile.
